From 726051e973d37e9d80956d6d1354f892b5fbe6dc Mon Sep 17 00:00:00 2001
From: Yongzhi Chen <ychen@cloudera.com>
Date: Thu, 12 Apr 2018 09:48:25 -0400
Subject: [PATCH 1349/1363] CLOUDERA-BUILD: CDH-67208 Select Struct failed
 with RuntimeException: Hive internal error:
 conversion of void to void not supported yet.
 Partial backport HIVE-18416 Initial support for
 TABLE function

Change-Id: Id5df010734b3e07887ac18b452cc0d04b45c5757
(cherry picked from commit 0145f167fcbf16b3b610df3da2db74745b457493)
---
 ql/src/test/queries/clientpositive/structin.q      |    2 ++
 ql/src/test/results/clientpositive/structin.q.out  |    9 +++++++++
 .../objectinspector/ObjectInspectorConverters.java |    6 ++++++
 3 files changed, 17 insertions(+)

diff --git a/ql/src/test/queries/clientpositive/structin.q b/ql/src/test/queries/clientpositive/structin.q
index 48b31f3..b64d633 100644
--- a/ql/src/test/queries/clientpositive/structin.q
+++ b/ql/src/test/queries/clientpositive/structin.q
@@ -15,3 +15,5 @@ struct('1234-1111-0074019112','1'),
 struct('1234-1111-0074019610','1'),
 struct('1234-1111-0074022106','1')
 );
+set hive.optimize.constant.propagation=true;
+select struct("x", null);
diff --git a/ql/src/test/results/clientpositive/structin.q.out b/ql/src/test/results/clientpositive/structin.q.out
index 81c792a..db30d6e 100644
--- a/ql/src/test/results/clientpositive/structin.q.out
+++ b/ql/src/test/results/clientpositive/structin.q.out
@@ -64,3 +64,12 @@ STAGE PLANS:
       Processor Tree:
         ListSink
 
+PREHOOK: query: select struct("x", null)
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+POSTHOOK: query: select struct("x", null)
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+{"col1":"x","col2":null}
diff --git a/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ObjectInspectorConverters.java b/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ObjectInspectorConverters.java
index 308373d..52d0a13 100644
--- a/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ObjectInspectorConverters.java
+++ b/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ObjectInspectorConverters.java
@@ -73,6 +73,12 @@ public Object convert(Object input) {
   private static Converter getConverter(PrimitiveObjectInspector inputOI,
       PrimitiveObjectInspector outputOI) {
     switch (outputOI.getPrimitiveCategory()) {
+    case VOID:
+      if (!outputOI.getTypeInfo().equals(inputOI.getTypeInfo())) {
+        throw new RuntimeException("Hive internal error: conversion of "
+            + inputOI.getTypeName() + " to void not possible.");
+      }
+      return new IdentityConverter();
     case BOOLEAN:
       return new PrimitiveObjectInspectorConverter.BooleanConverter(
           inputOI,
-- 
1.7.9.5

